### DISCUSSION TITLE
error 35 when upgrading from 8.8.0 to 8.10.1 (msys2/ucrt64)

### DISCUSSION THREAD

**Discussion by ReenigneArcher** [curl/curl#15099] [Q&A]
When upgrading from curl 8.8.0 to 8.10.1, we started getting error code 35 in our unit tests. It's a 100% failure. We have the following function to download a file:

```cpp
  bool download_file(const std::string &url, const std::string &file) {
    CURL *curl = curl_easy_init();
    if (curl) {
      // sonar complains about weak ssl and tls versions
      // ideally, the setopts should go after the early returns; however sonar cannot detect the fix
      curl_easy_setopt(curl, CURLOPT_SSLVERSION, CURL_SSLVERSION_TLSv1_2);
    }
    else {
      BOOST_LOG(error) << "Couldn't create CURL instance";
      return false;
    }

    std::string file_dir = file_handler::get_parent_directory(file);
    if (!file_handler::make_directory(file_dir)) {
      BOOST_LOG(error) << "Couldn't create directory ["sv << file_dir << ']';
      curl_easy_cleanup(curl);
      return false;
    }

    FILE *fp = fopen(file.c_str(), "wb");
    if (!fp) {
      BOOST_LOG(error) << "Couldn't open ["sv << file << ']';
      curl_easy_cleanup(curl);
      return false;
    }

    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, fwrite);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, fp);
#ifdef _WIN32
    curl_easy_setopt(curl, CURLOPT_SSL_OPTIONS, CURLSSLOPT_NATIVE_CA);
#endif

    CURLcode result = curl_easy_perform(curl);
    if (result != CURLE_OK) {
      BOOST_LOG(error) << "Couldn't download ["sv << url << ", code:" << result << ']';
    }

    curl_easy_cleanup(curl);
    fclose(fp);
    return result == CURLE_OK;
  }
```

result:
```
[ RUN      ] DownloadFileTests/DownloadFileTest.Run/0
D:/a/Sunshine/Sunshine/tests/unit/test_httpcommon.cpp:45: Failure
Value of: http::download_file(url, path)
  Actual: false
Expected: true

[2024-09-30 13:25:39.509]: Tests: From D:/a/Sunshine/Sunshine/tests/unit/test_httpcommon.cpp:41
[2024-09-30 13:25:39.509]: Tests:   DownloadFileTests/DownloadFileTest/Run/0 started
[2024-09-30 13:25:39.629]: Error: Couldn't download [https://httpbin.org/base64/aGVsbG8h, code:35]
[2024-09-30 13:25:39.629]: Tests: At D:/a/Sunshine/Sunshine/tests/unit/test_httpcommon.cpp:45
[2024-09-30 13:25:39.629]: Tests:   Failure: Value of: http::download_file(url, path)
  Actual: false
Expected: true
[2024-09-30 13:25:39.629]: Tests: DownloadFileTests/DownloadFileTest/Run/0 failed
[  FAILED  ] DownloadFileTests/DownloadFileTest.Run/0, where GetParam() = ("https://httpbin.org/base64/aGVsbG8h", "hello.txt") (119 ms)
[ RUN      ] DownloadFileTests/DownloadFileTest.Run/1
D:/a/Sunshine/Sunshine/tests/unit/test_httpcommon.cpp:45: Failure
Value of: http::download_file(url, path)
  Actual: false
Expected: true

[2024-09-30 13:25:39.629]: Tests: From D:/a/Sunshine/Sunshine/tests/unit/test_httpcommon.cpp:41
[2024-09-30 13:25:39.629]: Tests:   DownloadFileTests/DownloadFileTest/Run/1 started
[2024-09-30 13:25:39.738]: Error: Couldn't download [https://httpbin.org/redirect-to?url=/base64/aGVsbG8h, code:35]
[2024-09-30 13:25:39.738]: Tests: At D:/a/Sunshine/Sunshine/tests/unit/test_httpcommon.cpp:45
[2024-09-30 13:25:39.738]: Tests:   Failure: Value of: http::download_file(url, path)
  Actual: false
Expected: true
[2024-09-30 13:25:39.738]: Tests: DownloadFileTests/DownloadFileTest/Run/1 failed
[  FAILED  ] DownloadFileTests/DownloadFileTest.Run/1, where GetParam() = ("https://httpbin.org/redirect-to?url=/base64/aGVsbG8h", "hello-redirect.txt") (108 ms)
[----------] 2 tests from DownloadFileTests/DownloadFileTest (228 ms total)
```

Any ideas how we can solve this?

Note: This is specifically with msys2 ucrt64 version of curl. Doesn't seem to be a global problem as our ArchLinux build is still working, and they are on curl 8.10.1 as well.

PR: https://github.com/LizardByte/Sunshine/pull/3246

**Comment by dfandrich** [comment]
Error 35 is "SSL connect error. The SSL handshaking failed."  What TLS library
is it using? Are the working and non-working versions using the same library?
What protocol?  At the same version?  Can you show the debug logs with
curl_global_trace() with "ssl" enabled?

**Reply by ReenigneArcher** [reply]
Thanks for the fast response!

> What TLS library is it using?

If I understand the question correctly, `mingw-w64-ucrt-x86_64-openssl`

> Are the working and non-working versions using the same library?

As far as I understand, yes... No other changes on our side, except curl version. I didn't see any noticeable change in https://github.com/msys2/MINGW-packages/commits/master/mingw-w64-curl (but I don't really know what to look for specifically).

> What protocol?  At the same version?

TLS 1.2. It's specifically defined in the code snip. The server doesn't support 1.3. -> https://www.ssllabs.com/ssltest/analyze.html?d=httpbin.org&s=44.194.234.203&latest

> Can you show the debug logs with curl_global_trace() with "ssl" enabled?

I will have to report back on that.

**Reply by ReenigneArcher** [reply]
Seems to work when using `mingw-w64-ucrt-x86_64-curl-winssl` instead of `mingw-w64-ucrt-x86_64-openssl`.

**Reply by ReenigneArcher** [reply]
I mean replacing `mingw-w64-ucrt-x86_64-curl` with `mingw-w64-ucrt-x86_64-curl-winssl`.

**Comment by dfandrich** [comment]
> If I understand the question correctly, mingw-w64-ucrt-x86_64-openssl

What version of OpenSSL?

>     Are the working and non-working versions using the same library?
> 
> As far as I understand, yes... No other changes on our side, except curl
> version. I didn't see any noticeable change in https://github.com/msys2/
> MINGW-packages/commits/master/mingw-w64-curl (but I don't really know what to
> look for specifically).

Calling curl_version_info() is the surest way to find out. Sometime the version
curl was compiled with ends up being run with a different version of DLL at
run-time which really confuses things. Comparing the working and non-working
versions; you might be surprised that they're different.

>     What protocol? At the same version?
> 
> TLS 1.2. It's specifically defined in the code snip. The server doesn't support
> 1.3. -> https://www.ssllabs.com/ssltest/analyze.html?d=httpbin.org&s=
> 44.194.234.203&latest

Is this with HTTP/1.1, HTTP/2 or HTTP/3?
